packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -1006,21 +1027,26 @@ class _RenderDecoration extends RenderBox {
       ? maxContainerHeight
       : math.min(contentHeight, maxContainerHeight);
 
-    // Always position the prefix/suffix in the same place (baseline).
+    // Try to consider the prefix/suffix as part of the text when aligning it.
+    // If the prefix/suffix overflows however, allow it to extend outside of the
+    // input and align the remaining part of the text and prefix/suffix.
     final double overflow = math.max(0, contentHeight - maxContainerHeight);
-    final double baselineAdjustment = fixAboveInput - overflow;
+    final double textAlignVerticalFactor = (textAlignVertical.y + 1.0) / 2.0;
+    final double baselineAdjustment = fixAboveInput - overflow * (1 - textAlignVerticalFactor);
 
     // The baselines that will be used to draw the actual input text content.
-    final double inputBaseline = contentPadding.top
+    final double topInputBaseline = contentPadding.top
       + topHeight
       + inputInternalBaseline
       + baselineAdjustment;
-    // The text in the input when an outline border is present is centered
-    // within the container less 2.0 dps at the top to account for the vertical
-    // space occupied by the floating label.
-    final double outlineBaseline = inputInternalBaseline
```

Comment:
outlineBaseline is no longer needed because centering is done by textAlignVertical now.

Reply:
Maybe some of this comment needs to be retained above?